### ISSUE TITLE
Undefined index in user_menu_alter() 

### ISSUE THREAD

**Issue by stpaultim** [backdrop/backdrop-issues#5787]
## Description of the bug

We are periodically getting this error on a site. We can't reproduce the error reliably, but it continues to appear on a frequent basis.

Notice: Undefined index: admin/config/people/manage/fields in user_menu_alter() (line 1568 of /var/www/---/prod/src/core/modules/user/user.module).

## Steps To Reproduce

So after additional experimentation, I can reliably trigger the error by clearing the cache. Anytime I clear the cache and then visit a page, I trigger the notice. Clearing the cache does not trigger the error. Visiting a page after clearing the cache does.

This seems weird, but as far as I can tell. Visiting a Basic Page will trigger the notice on on the actual page. I only seem to get this message once per clearing of the cache. There is no watchdog log notice to go with this.

![image](https://user-images.githubusercontent.com/3144571/191102297-75e5dbba-b1c0-434d-a3fd-eb26c040decb.png)

The first Taxonomy page I visit after clearing the cache, triggers an error in the watchdog log, but no message on the screen. 

Each time I clear the cache, I seem to get one on screen message for visits to nodes AND one watchdog notice for visiting taxonomy pages. But, only one of each. 

## Other information

This site is a Drupal 7 to Backdrop upgrade with quite a few contributed modules enabled.

![image](https://user-images.githubusercontent.com/3144571/191099717-550ef457-fae6-4c41-9dda-5a4daa48fcda.png)

## List of all contrib modules on this site
<details>
<summary>Click to expand:</summary>

- field_wrappers 1.x-1.1.2
- file_entity_inline 1.x-1.0.0-beta1
- font_awesome 1.x-1.2.6
- google_cse 1.x-3.0.1
- google_tag 1.x-1.6.0
- googleanalytics 1.x-2.6.2
- imce 1.x-1.0.0
- inline_entity_form 1.x-1.9.1
- job_scheduler 1.x-1.1.4
- jstorage 1.x-1.0.1
- Paragraphs 1.x-1.2.0 
- Automatic Nodetitles 1.x-1.0.4
- Backup and Migrate 1.x-1.0.20
- Better Formats 1.x-1.0
- Block reference 1.x-2.4.0
- CAPTCHA 1.x-1.7.2
- Includes: CAPTCHA, Image CAPTCHA
- Colorbox 1.x-2.17.1
- Content view access 1.x-1.0.1
- Dashboard Plus 1.x-1.0.3
- Devel 1.x-1.8.1
- Disable Term Content Listings 1.x-1.0.5
- Draggableviews 1.x-1.0.1
- Includes: Draggableviews, Draggableviews Book
- Editable Fields 1.x-1.0.1
- Entity Plus 1.x-1.0.17
- Entity UI 1.x-1.0.1
- External Links 1.x-1.18.2
- Feeds 1.x-1.1.4
- Includes: Feeds, Feeds Admin UI, Feeds Import
- Libraries 1.x-2.2.2
- Markup 1.x-1.0.0
- Masquerade 1.x-1.0.0
- Maxlength 1.x-3.3.1
- Metatag 1.x-1.22.2
- Mobile Detect 1.x-1.1.1
- Node clone 1.x-1.1.4
- Node Clone Tab 1.x-1.1.0
- On The Web 1.x-1.7.2
- Paragraphs jQuery UI Accordion 1.x-1.0.1
- Publication Date 1.x-2.2.0
- reCAPTCHA 1.x-2.2.6
- SMTP Authentication Support 1.x-1.3.4
- Special Menu Items 1.x-2.1.3
- Splashify 1.x-1.3.4
- User Guide Recipe 1.x-1.0.2
- Video Filter 1.x-3.4.1
- Viewfield 1.x-1.0.0
- Views Conditional 1.x-1.3.1
- Webform 1.x-4.24.2
- Weight 1.x-1.0.2
</details>

**Comment by indigoxela** [comment]
@stpaultim as this is hard to reproduce on a vanilla install, it might be helpful to provide some more info about the site that shows the notices.

Usual suspects:

- Upgraded from Drupal or fresh B install
- List of enabled contrib modules
- Other, at first sight unrelated notices/warnings in dblog

Most of all: the **complete** info from dblog, including the page where it was triggered and the referrer. (Was it cron? Was it a specific page?...)

**Comment by stpaultim** [comment]
I realize that this wasn't enough information to do much yet. My colleague @rajpaul plans to add a little information about what he thinks is happening here. I'll update the original issue with a little more info. We'll definitely be helping to debug the issue, but wanted to create a stub issue in case anyone else has noticed anything.

**Comment by stpaultim** [comment]
I have had some luck reliably reproducing the problem. I've updated the original post with this new information.

**Comment by indigoxela** [comment]
Interesting, many thanks for the additional info. @stpaultim what do you see on /admin/config/people/manage and /admin/config/people/manage/fields?
Directly after flushing caches and after that.

That this only happens when flushing caches seems reasonable - hook_menu / hook_menu_alter get cached pretty well.

**Comment by stpaultim** [comment]
Sorry, my colleague who was planning to help with this issue is dealing with a family emergency. 

I'm not getting any special notices on the pages or in the log for the two paths you asked about, after clearing the cache. 

-  /admin/config/people/manage
- /admin/config/people/manage/fields

**Comment by indigoxela** [comment]
No problem - family first. :family: 

You'll probably have to do some more digging to figure out what's causing that "undefined index" in hook_menu_alter. Let us know, what you discover.

**Comment by argiepiano** [comment]
I'm not able to reproduce this issue. My guess is that this may be caused by a custom or contrib module that is messing with the menu items before reaching the user alter implementation. @stpaultim you may want to post a list of contrib modules in use.

**Comment by kiamlaluno** [comment]
I agree: It must be a contributed/custom module altering the routes before `user_menu_alter()`. 

Still, `user_menu_alter()` should first check the route it wants to alter exist, as the only way to remove a route, which is something contributed modules can do, is removing the route definition.

**Comment by rajpaul** [comment]
I checked this issue and found that `user_menu_alter` assigns an array item without checking it existence. I also checked if there are other contrib/custom modules use `hook_menu_alter` and used same array item but not. So I assume if we just can check if the menu item is exist or  not and then assign to the menu item array then it should be fixed. I think the below bold line causes the issue without checking null - 

`/**
 * Implements hook_menu_alter().
 */
function user_menu_alter(&$items) {
  if (module_exists('field_ui')) {
    // Provide a parent menu item for managing fields.
    **$items['admin/config/people/manage'] = $items['admin/config/people/manage/fields'];**
    $items['admin/config/people/manage']['type'] = MENU_NORMAL_ITEM;
    $items['admin/config/people/manage']['weight'] = 1;
    $items['admin/config/people/manage']['description'] = 'Configure fields and display of fields on user accounts.';

    // Set "Manage fields" tab as the default tab.
    $items['admin/config/people/manage/fields']['type'] = MENU_DEFAULT_LOCAL_TASK;
  }
}`

**Comment by indigoxela** [comment]
@rajpaul many thanks for trying to figure out what's going on. :pray: 

I do have concerns to just change the core code as proposed:

1. Nobody can reproduce the problem, there's a single site that seems to have it
2. Catching it in user_menu_alter() would hide the symptom on your site, but won't fix the actual problem (IMO)

We really need steps to reproduce.

**Comment by stpaultim** [comment]
> We really need steps to reproduce.

Well, we've shared our experiences and @rajpaul has shared his ideas on what might be going on. I undestand that this might not be enough to work with, but at least we have an issue opened in case someone else has a similar problem. 

I had asked @rajpaul to submit a PR, but @indigoxela seems to be indicating that wouldn't be useful at this point without steps that someone else can reproduce, so we'll hold off until we have more solid information.

**Comment by indigoxela** [comment]
@stpaultim many thanks for your patience.

There's still one thing you could provide: a list of enabled modules. That would make it easier to narrow down the root cause, in case someone else comes up with the same Notice - by just comparing the modules the sites share.

**Comment by stpaultim** [comment]
I have added a list of enabled contrib modules in a collapsed list (it's quite long) in the original post.

**Comment by kiamlaluno** [comment]
@stpaultim May you check there are `hook_menu_alter()` implementations from the enabled modules?

**Comment by argiepiano** [comment]
I've checked all the contrib modules you posted. As far as I can tell none of them alters or removes that particular path. Only one or two implement that hook. So the issue is somewhere else - hard to tell where.

Regarding the idea of wrapping this with an if statement as @rajpaul suggested: despite @indigoxela's reservations about this, I agree with @kiamlaluno that it may actually be the right way to approach a hook_menu_alter implementation, since it's possible that other contrib modules would in fact remove a route before the implementation is run (though I'm not saying this is the case here). See for example the hook_mnenu_alter implementation in the contrib [filefield_paths](https://github.com/backdrop-contrib/filefield_paths/blob/a778ab8f7b1dd04118bc7a61c8f6420f2d1e865c/includes/image.inc#L26).
